File: src/main/java/core/scm/specialforms/Future.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.exceptions.IllegalSyntaxException;
import core.scm.Cons;

import java.util.List;

/* Syntax:
 * (future <expression>)
 */
public enum Future implements ISpecialForm {
  FUTURE;

  @Override
  public Object eval(List<Object> expression, Environment env, Evaluator evaluator) {
    if (expression.size() < 2) {
      throw IllegalSyntaxException.of(toString(), expression);
    }
    core.scm.Future future;
    if (expression.size() > 2) {
      Cons list = Cons.list(Begin.BEGIN);
      list.addAll(expression.subList(1, expression.size()));
      future = new core.scm.Future(list, env, evaluator);
    } else {
      future = new core.scm.Future(expression.get(1), env, evaluator);
    }
    Evaluator.executor.submit(future);
    return future;
  }

  @Override
  public String toString() {
    return "future";
  }
}

File: src/main/java/core/scm/specialforms/Future.kt
package core.scm.specialforms

import core.environment.Environment
import core.evaluator.Evaluator
import core.exceptions.IllegalSyntaxException
import core.scm.Cons

/* Syntax:
 * (future <expression>)
 */
object Future : SpecialForm("future") {

    override fun eval(form: List<Any?>, env: Environment, evaluator: Evaluator) = when (form.size) {
        1    -> throw IllegalSyntaxException(toString(), form)
        2    -> form[1]
        else -> Cons.list<Any?>(Begin).apply { addAll(form.subList(1, form.size)) }
    }.let { core.scm.Future(it, env, evaluator).apply { Evaluator.executor.submit(this) } }
}
